gh-132983: Introduce `_zstd` bindings module

emmatyping opened the pull request:
This is part 2 (but parallel to part 1) for implementing PEP 784. This is probably the meat of the implementation making up ~half of the LOC in the overall diff.

This change:
- Adds the pyzstd license, since this is the first code introduced based on pyzstd
- Adds the code for the `_zstd` module in `Modules/_zstd`. Tests will be included when the Python `compression.zstd` module is added. If people think the tests should be included, I can merge the `compression.zstd` changes into the PR, but I wanted to separate them to make the PRs a bit more manageable.
- Adds unix build changes (Windows requires changes in cpython-source-deps and will be added in a follow-up PR)

I added skip news as I'd like to write a holistic NEWS/What's New entry once the entire implementation has landed. If people think each PR should have NEWS I can write something up.

TODO before merge:
- [x] try running tests on top of this branch
- [x] format PEP 7 style
- [x] run buildbots on this PR
- [x] verify `_zstd` gets installed

<!-- gh-issue-number: gh-132983 -->
* Issue: gh-132983
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--133027.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

emmatyping committed:
Add _zstd module

This commit introduces the `_zstd` module, with bindings to libzstd from
the pyzstd project. It also includes the unix build system configuration.
Windows build system support will be integrated independently as it
depends on integration with cpython-source-deps.

emmatyping committed:
Add _zstd to modules

emmatyping committed:
Fix path for compression.zstd module

ZeroIntensity commented on Modules/_zstd/zdict.c:
```diff
@@ -0,0 +1,304 @@
+    Py_XDECREF(self->dict_content);
+
+    /* Thread lock */
+    if (PyMutex_IsLocked(&self->lock)) {
```
This looks racy. If this thread doesn't own the lock, then another thread should release it. Either this call or the holding thread's call with cause a fatal error if it's locked.

gpshead replied:
Realistically this is a dealloc function.  by the time it is called in a gc destructor context, there shouldn't be anything else that _could_ unlock it.

ex: if we added an assert(!PyMutex_IsLocked(...)) in here could it _ever_ fail in a situation that isn't a fundamental bug elsewhere?

ZeroIntensity commented on Modules/_zstd/_zstdmodule.h:
```diff
@@ -0,0 +1,245 @@
+    PyObject_HEAD
+
+    /* Thread lock for generating ZSTD_CDict/ZSTD_DDict */
+    PyMutex lock;
```
In general, we don't add locks to object structures. We use the object's `ob_lock` via a critical section.

ZeroIntensity commented on Modules/_zstd/_zstdmodule.h:
```diff
@@ -0,0 +1,245 @@
+    /* __init__ has been called, 0 or 1. */
+    int inited;
+
+    _zstd_state *module_state;
```
I'm not familiar with this practice. Generally, you use `PyModule_GetState`. Are you following another module for precedent?

gpshead replied:
FYI - best practices or cpython stdlib code style issues can be worked out as followups after beta1 if any of these are deemed a challenge to tackle under time pressure for beta1.  Comments that cannot be addressed here can be tracked as new issues or a checklist of items in a new issue once this goes in.  I'll leave that up to Emma to decide what to tackle now vs later.

Please realize this is code coming from upstream project with a few authors over the last several years so i'd phrase review questions as "Is this ..." rather than "Are you ..." as that focuses on the more on the broader code at hand.

emmatyping replied:
Yeah I think this code pre-dates `PyType_GetModuleState` so it was easier for the original author to store the state in the objects.

I removed this in 388b69f as well as expanding out a lot of the module state macros.

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,660 @@
+    if (PyLong_Check(level_or_options)) {
+        const int level = PyLong_AsInt(level_or_options);
+        if (level == -1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
```
Seems like you want `OverflowError`.

gpshead replied:
ValueError is preferred for this, overflow is supposed to be "during math" rather than input value out of range.

ZeroIntensity commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,1125 @@
+
+    sizes_sum = 0;
+    for (i = 0; i < chunks_number; i++) {
+        PyObject *size = PyList_GET_ITEM(samples_size_list, i);
```
This isn't thread-safe. You need to use `PyList_GetItemRef`.

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,660 @@
+    PyObject *capsule;
+    ZSTD_CDict *cdict;
+
+    ACQUIRE_LOCK(self);
```
This isn't safe. `PyDict_GetItemWithError`, `PyLong_FromLong`, `Py_DECREF`, and a bunch of other things here can call back into Python. We need to use a critical section to prevent lock-ordering and re-entrancy deadlocks.

ZeroIntensity reviewed (commented):
Nice! Some comments at a glance.

Please remove the module state macros. They make this really difficult to review, and don't really add much (is `MS_MEMBER(whatever)` *really* that much easier to type than `state->whatever`?)

emmatyping committed:
Ignore _zstd module like _io

emmatyping committed:
Expand module state macros to improve code quality

Also removes module state references from the classes in the _zstd
module and instead uses PyType_GetModuleState()

StanFromIreland commented on Doc/license.rst:
```diff
@@ -1132,3 +1132,39 @@ The file is distributed under the 2-Clause BSD License::
+------------------
+
+Zstandard bindings in :file:`Modules/_zstd` and :file:`Lib/compression/zstd`
+are based on code from the pyzstd library, copyright Ma Lin and contributors.
```
Please add a link to the library.

StanFromIreland commented on Modules/_zstd/_zstdmodule.h:
```diff
@@ -0,0 +1,214 @@
+#pragma once
+/*
+Low level interface to Meta's zstd library for use in the `zstd` Python library.
```
```suggestion
Low level interface to Meta's zstd library for use in the zstd Python library.
```
No need for this in comments

StanFromIreland commented on Modules/_zstd/_zstdmodule.h:
```diff
@@ -0,0 +1,214 @@
+
+Original implementation by Ma Lin with patches by Rogdham.
+
+Refactored for the CPython standard library by Emma Harper Smith.
```
We have generally stopped updating these but I don't know about new ones. With *blame* there is not much need.

emmatyping replied:
I think it is nice to credit where code comes from (especially since blame will only show me), but I am fine removing this if that is the convention.

AA-Turner replied:
> especially since blame will only show me

Perhaps add git trailers to credit Lin & Rodgham? `Originally-Authored-By`/`Authored-By`/etc.

StanFromIreland replied:
I think the originally part we can leave but the new part should go.

adding them to the commit wouldn’t hurt either.

emmatyping replied:
I'll leave the module description and remove the rest. Perhaps when merged this PR can include the git trailers.

emmatyping replied:
Okay these should all be updated now.

StanFromIreland commented on Modules/_zstd/zdict.c:
```diff
@@ -0,0 +1,301 @@
+/*
+Low level interface to Meta's zstd library for use in the `zstd` Python library.
```
```suggestion
Low level interface to Meta's zstd library for use in the zstd Python library.
```
ditto

StanFromIreland commented on Modules/_zstd/zdict.c:
```diff
@@ -0,0 +1,301 @@
+
+Original implementation by Ma Lin with patches by Rogdham.
+
+Refactored for the CPython standard library by Emma Harper Smith.
```
ditto

StanFromIreland commented on Modules/_zstd/zdict.c:
```diff
@@ -0,0 +1,301 @@
+/* -----------------
+     ZstdDict code
+   ----------------- */
+
```
```suggestion
```
Why? Is the file name and previous comment not clear enough?

emmatyping replied:
This is an artifact from the existing code, it can probably be removed. I think it can be deferred until after this is merged in the interest of landing this before b1

picnixz replied:
I see your point. You can defer my comments as well. There are other comments that I didn't write yet that are not critical per se so it's fine to address them later

AA-Turner replied:
We shouldn't rush the implementation, though, beta 1 is still over a week away.

picnixz replied:
I'm actually leaving for the next 10 days tomorrow evening so I won't be able to reply. I'll try to finish the review tonight and you can decide what to ignore

emmatyping replied:
I definitely don't want to rush the implementation. There's a lot of code to consider and review, and I want to make sure it is well reviewed. There is another ~5k lines that need to be merged on top of this change, so I want to make sure both have time to land in time for the beta. I have been resolving  some of the more correctness/semantic concerns raised, and taking a few of the opinionated ones as well.

I think a lot of the review from @picnixz is semantic and/or covers correctness that should definitely be considered as part of the PR. Style things like removing comments seem less urgent to address. At least that is my two cents.

emmatyping replied:
I removed the repetitive comments in the most recent push.

emmatyping committed:
Remove backticks suggested in review

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,1151 @@
+    default:
+        Py_UNREACHABLE();
+    }
+    PyOS_snprintf(buf, sizeof(buf), msg, ZSTD_getErrorName(zstd_ret));
```
Use `PyErr_Format` instead of using `PyOS_snprintf()`.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,1151 @@
+    }
+
+    /* Prepare chunk_sizes */
+    chunk_sizes = PyMem_Malloc(chunks_number * sizeof(size_t));
```
You can use [`PyMem_New`](https://docs.python.org/3/c-api/memory.html#c.PyMem_New)  so that it takes care of a possible overflow.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,1151 @@
+                          int compression_level)
+/*[clinic end generated code: output=9c2a7d8c845cee93 input=ef6f4f2b0b7e178d]*/
+{
+    if (ZSTD_versionNumber() < 10405) {
```
Can we use a macro for 10405 instead of a magic number please? TiA

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,1151 @@
+        bound = ZSTD_dParam_getBounds(parameter);
+        if (ZSTD_isError(bound.error)) {
+            _zstd_state* const _module_state = PyModule_GetState(module);
+            if (_module_state != NULL) {
```
A NULL state is not good at all, we should abort in a DEBUG build.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,1151 @@
+    _zstd_state* const _module_state = PyModule_GetState(module);
+    if (_module_state == NULL) {
+        return NULL;
+    }
```
Let's have a function for recovering the module state (see other modules)

picnixz reviewed (commented):
Few preliminary comments, although I'm not sure it can be modified under the given license.

emmatyping committed:
Use critical sections to lock object state

This should avoid races and deadlocks.

emmatyping committed:
Remove compress/decompress and mark module as not reliant on the GIL

The `compress`/`decompress` functions will be moved to Python code for simplicity.
C implementations can always be re-added in the future.

Also, mark _zstd as not requiring the GIL.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+}
+
+typedef struct {
+    const int parameter;
```
General comment: we don't use `const T` when `T` is a primitive type such as int even to convene the intent of a constant variable.

emmatyping replied:
Is this true in argument positions too? e.g. should I remove the `const Py_ssize_t` from buffer.h?

picnixz replied:
I think we have a bit of both.

* Benefits:
  - avoid mistakes
  - convene the intent of a read-only value
* Downsides:
  - adds more words to read
  - for a function parameter, it does nothing except for a pointer where it makes sense (it helps in its usage but it doesn't help more the compiler)

It's really up to you though, but I would say that it'd be good to be consistent in the `const` usage. I however haven't looked in all places to see if structures are declared with a `const` or not.

But we can discuss this in a separate PR.

emmatyping replied:
Thanks for the context. I will think about it. If the tradition is to not have it in functions or structs then I think it would be good to remove them throughout.

emmatyping replied:
I removed consts on simple types.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+        list = cp_list;
+        list_size = Py_ARRAY_LENGTH(cp_list);
+        type = "compression";
+    } else {
```
Nit, but can you (in a follow-up PR if you want to) make it so that the new code follow PEP-7? this includes writing something like:

```c
if (...) {
...
}
else {
...
}
```

emmatyping replied:
I plan on making the code PEP 7 compliant prior to merge, but want to work through the semantic/functional issues first.

emmatyping replied:
I've fixed this but considering the timing I'll work on a more thorough review of style to make the code PEP 7 compliant once everything is landed.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+/*[clinic input]
+_zstd._train_dict
+
+        samples_bytes: PyBytesObject
```
Can you use 4-space indents here instead of 8 spaces in general? TiA.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+
+    /* Check zstd dict error */
+    if (ZDICT_isError(zstd_ret)) {
+        _zstd_state* const _module_state = PyModule_GetState(module);
```
if the module state is NULL, no exception is set. We should abort here. This is why it's preferrable to have a function, say, `get_module_state(module)` which returns the module state but with an assert. Something like:

```c
static _zstd_state *get_state(PyObject *m) {
    void *s = PyModule_GetState(m);
    assert(s != NULL);
    return (_zstd_state *)s;
}
```

emmatyping replied:
I pushed a commit adding a module get state function right after your review 😁

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+    if (dst_dict_bytes == NULL) {
+        goto error;
+    }
+
```
From the beginning of the function to this line, it appears that it's the same code as in the train() routine above. Can the logic be refactored into a single function?

emmatyping replied:
I think I will leave refactoring this and the instance below to a future PR. I'll add TODOs and try to address this after b1.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+                        PyBytes_AS_STRING(samples_bytes), chunk_sizes,
+                        (uint32_t)chunks_number, params);
+    Py_END_ALLOW_THREADS
+
```
And from here until the end, the logic is also the same. Can we also refactor this into a separate function?

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+                "complete frame. Zstd error message: %s.",
+                ZSTD_getErrorName(frame_size));
+        }
+        goto error;
```
Just return NULL here and then `return PyLong_FromSize_t(frame_size);`:

```c
frame_size = ....
if (ZSTD_isError(...)) {
   ...
   return NULL;
}
return PyLong_FromSize_t(frame_size);
```

emmatyping replied:
Much neater, thanks!

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+                "starts from the beginning of a frame, and its length "
+                "not less than the frame header (6~18 bytes).");
+        }
+        goto error;
```
Return `NULL` here and return `ret` at the end without labels:

```c
if (decompressed_size == ZSTD_CONTENTSIZE_ERROR) {
   ...
   return NULL;
}
dict_id = ...
if (decompressed_size == ZSTD_CONTENTSIZE_UNKNOWN) {
    return Py_BuildValue(...)
}
return Py_BuildValue(...)
```

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+                                PyObject *d_parameter_type)
+/*[clinic end generated code: output=a13d4890ccbd2873 input=3e7d0d37c3a1045a]*/
+{
+    _zstd_state* const _module_state = PyModule_GetState(module);
```
Let's rename it to "state". No need to make the variable "private" inside the function itself.

emmatyping replied:
I was thinking I'd do so, I agree the _module_state name makes no sense.

emmatyping replied:
I renamed it to `mod_state`. I felt that further into longer functions "state" was unclear what the state was *of*, so I wanted to include that it is the module state

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+
+    /* zstd_version_info, a tuple. */
+    obj = get_zstd_version_info();
+    if (PyModule_AddObject(module, "zstd_version_info", obj) < 0) {
```
Prefer using [`PyModule_AddObjectRef`](https://docs.python.org/3/c-api/module.html#c.PyModule_AddObjectRef) over `PyModule_AddObject`.

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+    }
+    Py_DECREF(temp);
+
+    return 0;
```
```suggestion
    int rc = PyObject_SetAttrString((PyObject*) type, name, temp);
    Py_DECREF(temp);
    return rc;
```

picnixz commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+static void
+_zstd_free(void *module)
+{
+    _zstd_clear((PyObject *)module);
```
```suggestion
    (void)_zstd_clear((PyObject *)module);
```

picnixz commented on Modules/_zstd/_zstdmodule.h:
```diff
@@ -0,0 +1,199 @@
+static inline _zstd_state *
+get_zstd_state(PyTypeObject *type) {
+    PyObject *module = PyType_GetModuleByDef(type, &_zstdmodule);
+    return PyModule_GetState(module);
```
```suggestion
    PyObject *module = PyType_GetModuleByDef(type, &_zstdmodule);
    if (module == NULL) {
        return NULL;
    }
    void *state = PyModule_GetState(module);
    assert(state != NULL);
    return (_zstd_state *)state;
```

picnixz commented on Modules/_zstd/buffer.h:
```diff
@@ -0,0 +1,105 @@
+_OutputBuffer_Grow(_BlocksOutputBuffer *buffer, ZSTD_outBuffer *ob)
+{
+    assert(ob->pos == ob->size);
+    Py_ssize_t res = _BlocksOutputBuffer_Grow(buffer, &ob->dst, 0);
```
`res` should be checked in case of an error.

picnixz commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,670 @@
+                return -1;
+            }
+
+            /* Both key & value should be 32-bit signed int */
```
Nit but if you want a 32-bit signed int, I suggest using `PyLong_AsInt32`, although I doubt this module will ever be part of an embedded Python on 16-bit archs.

emmatyping replied:
Hm, I *think* that the comment is a red herring and should be removed. The `ZSTD_CCtx_setParameter` function takes an `int`, not an `int32_t` and an enum value (so another `int`). Therefore I think the comment is wrong and should be updated.

emmatyping replied:
I removed this comment. In general I think it would be a better UX if we checked the bounds and report that rather than relying on the values  being `int32_t` or `int`. I left a TODO to add that.

picnixz commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,670 @@
+}
+
+static void
+ZstdCompressor_dealloc(ZstdCompressor *self)
```
Methods for PyObjects must always take a `PyObject *` as `self` and make the cast inside otherwise you'll get an "incorrect function signature" as it's an undefined behavior if we call this function with a `PyObject*` at runtime.

emmatyping committed:
Lift critical section to avoid clang warning

picnixz commented on Modules/_zstd/zdict.c:
```diff
@@ -0,0 +1,296 @@
+_zstd_ZstdDict___reduce___impl(ZstdDict *self)
+/*[clinic end generated code: output=5c9b8a3550429417 input=2d69a2b4775ff76d]*/
+{
+    PyErr_SetString(PyExc_TypeError,
```
I think it's better if we document this instead. If you still want it, shouldn't we prefer `NotImplementedError` over `TypeError`?

emmatyping replied:
I agree documenting this instead is probably the better solution.

picnixz commented on Modules/_zstd/zdict.c:
```diff
@@ -0,0 +1,296 @@
+                  "<ZstdDict dict_id=%u dict_size=%zd>",
+                  dict->dict_id, Py_SIZE(dict->dict_content));
+
+    return PyUnicode_FromString(buf);
```
Use `PyUnicode_Format`

emmatyping replied:
I used `PyUnicode_FromFormat`, which I presume is what you meant. Thank you for recommending all of these C APIs. I don't have  as much experience working directly with the C API so it's really helpful to have someone point out cleaner/better ways of doing things! I presume a lot of these things are the way they are because the extension was written several years ago.

picnixz reviewed (commented):
Some other comments I managed to find. I didn't review the decompressor but I think most of my comments that I said for the compressor would apply. The most important ones are essentially function signatures where I want you to use `PyObject *self` and do a manual cast to the expected type inside the function:

```c
#define MyType_CAST(op)	((MyType *)(op))

static void
my_method(PyObject *op) {
    MyType *self = MyType_CAST(op);
    ...
}
```

The use of a macro (or a static inline function, up to you) is to allow future runtime type checks.

AA-Turner commented on Doc/license.rst:
```diff
@@ -1132,3 +1132,39 @@ The file is distributed under the 2-Clause BSD License::
+
+Zstandard bindings in :file:`Modules/_zstd` and :file:`Lib/compression/zstd`
+are based on code from the pyzstd library, copyright Ma Lin and contributors.
+The code is distributed under the 3-Clause BSD License::
```
Minor, but because this isn't just a copy of the pyzstd work, we should be clear that it is only the original upstream that is BSD-3.
```suggestion
The ``pyzstd`` code is distributed under the 3-Clause BSD License::
```

AA-Turner commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+        if (chunk_sizes[i] == (size_t)-1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Items in samples_size_list should be an int "
+                            "object, with a size_t value.");
```
If this exception can be triggered by Python code, we should avoid using C types in it.

emmatyping replied:
I went through all of the exception messages and tried to make sure this was done consistently. Hopefully they look better now!

AA-Turner commented on Modules/_zstd/_zstdmodule.c:
```diff
@@ -0,0 +1,968 @@
+    chunks_number = Py_SIZE(samples_size_list);
+    if ((size_t) chunks_number > UINT32_MAX) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The number of samples should <= UINT32_MAX.");
```
Likewise, if this exception can be triggered by Python code, we should avoid using C types in it.

```suggestion
        PyErr_SetString(PyExc_ValueError,
                        "The number of samples must be less than 2**32.");
```

picnixz replied:
We have other places (e.g., `_pysqlite_set_result`) where we mention INT_MAX but other places where we use `%i` and `INT_MAX`. I suggest using `%u` to format `UINT32_MAX` (IIRC, it's `%u` for `uint32_t`).

AA-Turner replied:
Changing existing usages might be trickier, but we should try and avoid introducing new ones.

picnixz replied:
I actually myself introduced some when I implemented HMAC, so I'll amend this as well.

emmatyping replied:
I used a `%u` format with `UINT32_MAX` to present the range.

emmatyping committed:
Respond to comments by picnixz

emmatyping committed:
Call out pyzstd explicitly in license description

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>

emmatyping committed:
Use a much more robust implementation...

... for `get_zstd_state_from_type`

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

emmatyping committed:
Use PyList_GetItemRef for thread safety purposes

emmatyping committed:
Use a macro for the minimum supported version

emmatyping committed:
remove const from primivite types

emmatyping committed:
Use PyMem_New in another spot

emmatyping committed:
Simplify error handling in _get_frame_size

emmatyping committed:
Another simplification of error handling in get_frame_info

emmatyping committed:
Rename _module_state to mod_state

emmatyping committed:
Rewrite comment explaining the context of the code

emmatyping committed:
Add link to pyzstd

emmatyping committed:
Add TODO about refactoring dict training code

emmatyping committed:
Use PyModule_AddObjectRef over PyModule_AddObject

PyModule_AddObject is soft-deprecated, so we should use PyModule_AddObjectRef

emmatyping committed:
Check result of OutputBufferGrow

emmatyping committed:
Simplify return logic in `add_constant_to_type`

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

emmatyping committed:
Ignore return value of _zstd_clear()

Co-authored-by: Bénédikt Tran <10796600+picnixz@users.noreply.github.com>

emmatyping committed:
Remove redundant comments

emmatyping committed:
Remove __reduce__ from ZstdDict

We should instead document that to pickle a dictionary a user should use
the `.dict_content` attribute.

emmatyping committed:
Use PyUnicode_FromFormat instead of a buffer

emmatyping committed:
Don't use C constants/types in error messages

emmatyping committed:
Make error messages easier to understand for Python users

emmatyping committed:
Lower minimum required version 1.4.0

emmatyping committed:
Use casts and make slot function signatures correct

emmatyping committed:
Be consistent with CPython on const usage

emmatyping committed:
Make else clauses in line with PEP 7

emmatyping committed:
Fix over-indented blocks in argument clinic

emmatyping committed:
Merge branch 'main' into 3.14-zstd-c-code

emmatyping committed:
Merge branch 'main' into 3.14-zstd-c-code

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,681 @@
+    PyObject *capsule;
+    ZSTD_CDict *cdict;
+
+    Py_BEGIN_CRITICAL_SECTION(self);
```
I think it's OK here, but in general, we use the following convention for critical section functions:

```c
static whatever *
my_func_lock_held(PyObject *self, whatever *args)
{
    _Py_CRITICAL_SECTION_ASSERT_OBJECT_LOCKED(self);
    /* actual implementation */
}

whatever *
my_func(PyObject *self, whatever *args)
{
    whatever *res;
    Py_BEGIN_CRITICAL_SECTION(self);
    res = my_func_lock_held(self, args);
    Py_END_CRITICAL_SECTION();
    return res;
}
```

This avoids accidentally bailing out of the critical section early.

emmatyping replied:
I'll leave a TODO to handle this for post-b1.

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,681 @@
+
+        /* Create ZSTD_CDict instance */
+        Py_BEGIN_ALLOW_THREADS
+        cdict = ZSTD_createCDict(PyBytes_AS_STRING(self->dict_content),
```
It's unsafe to call `PyBytes_AS_STRING` or `Py_SIZE` inside a `Py_BEGIN_ALLOW_THREADS` block. On the default build, those fields are synchronized by the GIL, so this is racy.

emmatyping replied:
Wouldn't the dict contents be protected by the lock on self?

ZeroIntensity replied:
Nope. Critical sections are released inside `Py_BEGIN_ALLOW_THREADS`, and they don't exist at all on GIL-ful builds. As a general rule of thumb, you can't call any of the C API when the thread state is detached, unless the function explicitly documents otherwise.

emmatyping replied:
I've pushed [f02ff5a](https://github.com/python/cpython/pull/133027/commits/f02ff5ac0ee4d304c9267d77f3301e3935a2bff6) which should hopefully address this.

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,681 @@
+        zstd_ret = ZSTD_CCtx_loadDictionary(
+                            self->cctx,
+                            PyBytes_AS_STRING(zd->dict_content),
+                            Py_SIZE(zd->dict_content));
```
I'd rather use `PyBytes_GET_SIZE`, but it's not an actual issue right now. `Py_SIZE` is intentionally thread-unsafe for performance reasons, but it's not an issue because bytes are immutable anyway.

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,681 @@
+    }
+    else {
+        /* Impossible code path */
+        PyErr_SetString(PyExc_SystemError,
```
Use `Py_UNREACHABLE()`.

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,681 @@
+        goto error;
+    }
+
+    assert(self->dict == NULL);
```
I'm pretty sure these assertions aren't safe. `tp_alloc` doesn't zero-out the memory.

emmatyping replied:
Good point, I will remove these assertions.

emmatyping replied:
I checked the docs, and tp_alloc actually *does* zero out the memory:

> The purpose of this function is to separate memory allocation from memory initialization. It should return a pointer to a block of memory of adequate length for the instance, suitably aligned, and initialized to zeros

https://docs.python.org/3/c-api/typeobj.html#c.allocfunc

ZeroIntensity replied:
I'm like, 90% sure that's wrong. There's a big PR from several months ago that fixes these things, but it hasn't landed yet.

ZeroIntensity replied:
Double checked, the docs are wrong. `_PyObject_Malloc` does not zero-out memory.

emmatyping replied:
Thanks for checking, I removed these asserts.

ZeroIntensity commented on Modules/_zstd/compressor.c:
```diff
@@ -0,0 +1,681 @@
+    ZSTD_freeCCtx(self->cctx);
+
+    /* Py_XDECREF the dict after free the compression context */
+    Py_XDECREF(self->dict);
```
Shouldn't this be a garbage collected type? If you can hold a reference to a Python object, you need garbage collection in basically all cases.

emmatyping replied:
My understanding is that garbage collection is only for types where there could be reference cycles, which is not possible with any of these types. While these types refer to other objects, those objects are internal to the type, thus a reference cycle is impossible.

ZeroIntensity replied:
Yes, but you have to be very careful with it. If you use an instance of a pure-Python type anywhere here, then that will leak, because instances hold a reference to their type, and types come with reference cycles through their MRO.

emmatyping replied:
This reference is to a `ZstdDict` instance, so no issue there. The other references in the (de)compressor are similarly references to C-implemented types.

ZeroIntensity replied:
I guess we can leave it for now to make the beta freeze. In any other PR, I'd reject this :)

ZeroIntensity commented on Modules/_zstd/decompressor.c:
```diff
@@ -0,0 +1,858 @@
+        }
+
+        /* Set parameter to compression context */
+        zstd_ret = ZSTD_DCtx_setParameter(self->dctx, key_v, value_v);
```
I would think we need a critical section here.

ZeroIntensity commented on Modules/_zstd/decompressor.c:
```diff
@@ -0,0 +1,858 @@
+decompressor_reset_session(ZstdDecompressor *self,
+                           decompress_type type)
+{
+    /* Reset variables */
+    self->in_begin = 0;
+    self->in_end = 0;
+
+    if (type == TYPE_DECOMPRESSOR) {
+        Py_CLEAR(self->unused_data);
+    }
+
+    /* Reset variables in one operation */
+    self->needs_input = 1;
+    self->at_frame_edge = 1;
+    self->eof = 0;
+    self->_unused_char_for_align = 0;
+
+    /* Resetting session never fail */
+    ZSTD_DCtx_reset(self->dctx, ZSTD_reset_session_only);
```
None of this is atomic. Do we need to acquire the critical section?

emmatyping replied:
This should be (and currently is) called from code where a lock on self is held. Maybe I should rename it to match the pattern used to indicate that.

ZeroIntensity commented on Modules/_zstd/decompressor.c:
```diff
@@ -0,0 +1,858 @@
+    ZSTD_freeDCtx(self->dctx);
+
+    /* Py_XDECREF the dict after free decompression context */
+    Py_XDECREF(self->dict);
```
Yeah, again if we have references to Python objects then we need it to be garbage collected 99% of the time.

ZeroIntensity commented on Modules/_zstd/zdict.c:
```diff
@@ -0,0 +1,263 @@
+    assert(self->inited == 0);
+
+    /* ZSTD_CDict dict */
+    self->c_dicts = PyDict_New();
```
Yeah, same comments from the other types. I think this needs to be GC-tracked, and those assertions don't seem right.

ZeroIntensity reviewed (commented):
I did my best to avoid any formality issues. The two big issues that I'm seeing are:

- The types aren't garbage-collected. They will leak during reference cycles.
- There's a couple missing critical sections. There needs to be locks around readers, not just writers.

Let's try to make the beta freeze!

emmatyping committed:
Add critical section around ZSTD_DCtx_setParameter

emmatyping committed:
Add a TODO about refactoring critical sections

emmatyping committed:
Use Py_UNREACHABLE

emmatyping committed:
Move bytes operations out of Py_BEGIN_ALLOW_THREADS

emmatyping committed:
Add TODO about ensuring a lock is held

emmatyping committed:
Remove asserts that may not be correct

emmatyping committed:
Add TODO to make ZstdDict and others GC objects

emmatyping committed:
Make objects GC tracked

emmatyping committed:
Remove unused include

emmatyping committed:
Fix some memory issues

ZeroIntensity reviewed (approved):
LGTM, assuming we're going to deal with garbage collection in a follow-up (which might go after the beta freeze).

emmatyping committed:
Fix refleaks on module and in ZstdDict

emmatyping committed:
Update configure to check for ZDICT_finalizeDictionary

emmatyping committed:
Merge branch 'main' into 3.14-zstd-c-code

emmatyping committed:
Properly check version in configure

emmatyping committed:
exit(1) if check fails

emmatyping committed:
Use AC_RUN_IFELSE

emmatyping committed:
Use a define() to re-use version check

emmatyping committed:
Merge branch 'main' into 3.14-zstd-c-code

emmatyping committed:
Actually properly set _zstd module status based on version

emmatyping committed:
Merge branch 'main' into 3.14-zstd-c-code

emmatyping committed:
Merge branch 'main' into 3.14-zstd-c-code

ned-deily commented on configure.ac:
```diff
@@ -5415,6 +5415,40 @@ PKG_CHECK_MODULES([LIBLZMA], [liblzma], [have_liblzma=yes], [
 
+have_libzstd=no
+AC_DEFUN([TEST_ZSTD_VERSION],[
+  AC_MSG_CHECKING([if libzstd is new enough])
```
There seem to be two problems with the `new enough` check:
1. There should be `AC_MSG_RESULT([yes])` and `AC_MSG_RESULT([no])` calls to complete the messages:
```
checking for libzstd... yes
checking if libzstd is new enough... checking for hstrerror... yes

```

ned-deily commented on configure.ac:
```diff
@@ -5415,6 +5415,40 @@ PKG_CHECK_MODULES([LIBLZMA], [liblzma], [have_liblzma=yes], [
+  have_libzstd=yes
+])
+
+PKG_CHECK_MODULES([LIBZSTD], [libzstd], [TEST_ZSTD_VERSION()], [
```
and 2. There's something funky going on here with the two calls to `TEST_ZSTD_VERSION()` such that version number test fails and the failure is sort of hidden in `config.log`:
```
configure:22400: checking for libzstd
configure:22407: $PKG_CONFIG --exists --print-errors "libzstd"
configure:22410: $? = 0
configure:22424: $PKG_CONFIG --exists --print-errors "libzstd"
configure:22427: $? = 0
configure:22727: result: yes
configure:22730: checking if libzstd is new enough
configure:22756: gcc -o conftest    conftest.c -ldl  >&5
conftest.c:325:10: fatal error: 'zstd.h' file not found
  325 | #include "zstd.h"
      |          ^~~~~~~~
1 error generated.
configure:22756: $? = 1
configure: program exited with status 1

```

This later results in:
`checking for stdlib extension module _zstd... disabled
`

I suspect the problem is the call to `TEST_ZSTD_VERSION()` on line 5437 fails because at that point `CPPFLAGS` is no longer set to include `$LIBZSTD_CFLAGS` and the test compile would succeed only if the header files are installed in a default system location, like `/usr/include`, which won't be the case, for example, on builds on macOS.

ned-deily replied:
Postscript: a likely confirmation of my suspicion above. For our CI runs on macOS, we use Homebrew to supply third-party libs (like `libzstd`) not supplied by Apple. If you look at CI logs for current PRs (i.e. after this PR was merged), `_zstd` is disabled (as above) on the `ghcr.io/cirruslabs/macos-runner:sonoma `runner but succeeds on the `macos-13` runner.  The difference is that the former is Apple Silicon (arm64) hardware where Homebrew installs to a non-standard `/opt/homebrew` prefix, whereas the `macos-13` runner is Intel hardware where Homebrew continues to install to the legacy `/usr/local` prefix which, no doubt, is included in the compiler's default search path.

ned-deily reviewed (commented):
@emmatyping, @gpshead, @hugovk Some showstopper problems with trying to build `_zstd`.
